### ISSUE TITLE
Thoughts on using scala for writing tests in htsjdk?

### ISSUE THREAD

**Issue by tfenne** [samtools/htsjdk#781]
What do folks think about using scala for testing in htsjdk?  @yfarjoun and I were talking about this over lunch and thought this would be a good change overall.  @nh13 and I have been working a lot more in scala over the last couple of years, and one thing I've noticed is that unit tests (particularly using `scalatest`, tend to be much more concise and hence easier to write and easier to read and understand after the fact.

I'd be more than happy to spend a little time writing some tests in scala for htsjdk to get the ball rolling if others are on board with this.  The only thing I'd look for a little help with is getting the gradle configuration working - @lbergelson perhaps I could interest you in helping there?

It looks like the necessary pieces of the puzzle are there, including:

1. [Scala support](https://docs.gradle.org/current/userguide/scala_plugin.html) in gradle
1. A [scalatest plugin](https://github.com/maiflai/gradle-scalatest) for gradle
1. A way to make existing TestNG suites [runnable via scalatest](http://www.scalatest.org/getting_started_with_testng_in_scala)

**Comment by yfarjoun** [comment]
I agree that if it makes writing and reading tests easier, it's a Good
Thing (TM). I would want to be sure that otherwise things do not get worse.

So, to add upon the list you mentioned, I'd like to be sure that (these
might be trivially true, I just don't know):

1. The coverage reports (from CodeCov) include combined coverage from the
Java and scala tests,
2. It is easy (in IntelliJ for example) to navigate (via "usages", stack
traces etc.) to either languages
3. Single unit tests can still be run inside intelliJ as currently
4. It is accepted that no "Hard-core" scala is used, so that folks that
only know java are not barred from reading/improving scala tests.

**Comment by cmnbroad** [comment]
I'm not sure how many people we'd lose if we did that, but I personally would like to see it. Also, I think there are some areas where htsjdk would benefit from generative, property-based testing.  Plus it builds character.

I added a simple scalacheck property-based test, with basic Gradle integration [here](https://github.com/cmnbroad/htsjdk/commit/cbf79a47b6d727cce34e084171031be0f053c302). The IntelliJ debugging, cross language single-stepping, and test integration seems pretty good. I didn't try the scalatest Gradle plugin though (or regenerate the Gradle wrapper or anything).

As usual, the Scala compiler is slowwww...

**Comment by magicDGS** [comment]
How much difficult will be to implement/read the tests for a Scala-newb? On the other hand, will it be possible to still implement some tests in pure Java? I would like to continue contributing, but I don't know if I will be able to implement unit tests for my patches properly...

**Comment by tfenne** [comment]
@magicDGS: My take (although I'm biased) is that if you're familiar with Java 8's stream support that writing tests in scala, given a reasonable number as examples and sources of copy/paste, shouldn't be too bad.  This wouldn't preclude writing tests in Java, but I suspect we'd want to keep all unit tests for a given class either in Java or in Scala.

Here's a link to all the test sources in a scala project that @nh13 and I work on together that builds on htsjdk in a picard-like manner: https://github.com/fulcrumgenomics/fgbio/tree/master/src/test/scala/com/fulcrumgenomics

The following is a good example of a [simple test class.](https://github.com/fulcrumgenomics/fgbio/blob/master/src/test/scala/com/fulcrumgenomics/util/SequencesTest.scala)

**Comment by magicDGS** [comment]
Thanks for the explanation @tfenne! I will have a look, but if it's like Java 8's streams should be fine for me.

**Comment by yfarjoun** [comment]
I wonder if "we'd want to keep all unit tests for a given class either in
Java or in Scala" might be too strong. It would preclude adding tests in
scala to already existing classes, and would deter someone from starting a
good scala testing framework to an already existing class.

Why not allow XyzTest.java and XyZTest.scala in the same project (under the
appropriate directory)?

**Comment by tfenne** [comment]
@yfarjoun I just think it would likely get difficult to keep track of what tests a certain class has if there are both java and scala tests for it.  And your example of `XyzTest` and `XyZTest` is a good one - we couldn't even call both classes the same thing, so we'd have to invent artificial differences and that would cause yet more confusion.  My 2c, you may disagree.

**Comment by droazen** [comment]
I'm a little concerned about this proposal's potential implications for compilation times. @cmnbroad's test showed that compilation time more than doubled even with just one scala test, due to the slowness of the scala compiler.

**Comment by cmnbroad** [comment]
I think we need more observations; mine was purely anecdotal. What I noticed was a visible pause during scala compile. Having said that I just tried doing a clean rebuild of my branch a few times (on a different machine) and it seemed quite fast.

**Comment by tfenne** [comment]
The scala compiler is definitely slow-er, the question is just how much slower is tolerable.  I'm of the opinion that ~5-10s extra in compile time is worth it if it means writing tests is easier and faster, and as a result we get better test coverage.  Others may disagree.

**Comment by droazen** [comment]
I'd be in favor of some more empirical tests to try to figure out what the slowdown would be like.

**Comment by lbergelson** [comment]
Running `./gradlew compileTestJava` in @cmnbroad's branch takes about 8s on my machine, and running `./gradlew compileTestScala` (which also compiles the java code)  takes about 14. I tried adding a few more scala files and it went up to around 15 seconds. So, from this very naive testing, it looks like theres a few seconds constant cost + a fraction of a second per additional scala file.
That doesn't seem that worrying to me.

@tfenne in your projects is compile time a big issue for you?  It doesn't seem like it would be the deciding factor here.

A different question is is this really going to be tests only?  What about test infrastructure that downstream users would want to use?  The `SAMRecordSetBuilder` was one example that @yfarjoun gave as and example of something that could be greatly improved by rewriting in scala.   (by making use of default values instead of dozens of overloaded functions).  Are we going to have a scala version of it in our htsjdk artifact?  Would we offer a separate test infrastructure artifact?  Or are we going to just make htsjdk a mixed java/scala project.

**Comment by droazen** [comment]
@lbergelson Can you try compiling on, eg., a Broad server on an NFS filesystem to see whether the few seconds' difference you're seeing locally translates into additional minutes in a non-local dev environment?

Also a bit concerned that this change would make it difficult to determine what code is actually covered by tests. Adding new tests for an existing class would require additional time/mental overhead to check multiple locations in the source tree.

**Comment by yfarjoun** [comment]
this was one of my concerns (coverage) as well, and I suggested that we see
if it's possible to get a single codeCove report that includes all the
tests (java and scala)

**Comment by nh13** [comment]
@lbergelson we have some code elsewhere that wraps the `SamRecordSetBuilder` and adds `VariantContextSetBuilder`: https://github.com/fulcrumgenomics/fgbio/tree/master/src/main/scala/com/fulcrumgenomics/testing.  We find them incredibly useful.

**Comment by tfenne** [comment]
@lbergelson @droazen Coming back to the compile time - this hasn't really been an issue for me on projects where we have a lot of tests written in scala.  My experience is that I only compile the tests when I'm going to run them, and the test suite run time is many multiples of the compile time ... so a slightly longer compile time is not that bad.  Also, I suspect gradle can do this to, but sbt is very good at only rebuilding the tests that need rebuilding to cut down on compile time.

@yfarjoun Getting a single coverage report should be doable I think.  We have mixes of Scala and Java code in some places, and test them all with scala tests.  Scalatest can (with a tiny bit of retrofit) run TestNG tests so that they all run together.  I think all the pieces are there - we'd just have to test it.

**Comment by yfarjoun** [comment]
So would all the tests be run by scalatest so that the coverage includes
everything?

**Comment by tfenne** [comment]
All the tests would be run by scalatest.  I _think_ that would then lead to coverage being collected for all the tests, but we would need to check and make sure that actually works.

**Comment by tfenne** [comment]
Hi All.  I've created a PR, #789, that is a first pass at making this happen.  It was relatively painful for the following reasons:

1. Almost all the test classes in HTSJDK don't descend from anything except `Object`, and best way to make them runnable via scalatest _and_ TestNG is to parent them under a specific class that enables this.  I had to hunt down every test class and add the extends clause :(  Even if we don't start using scala (I think we should) this part of the PR should probably get merged as it will likely  be helpful in future.

2. Gradle, its DSL, and its plugin language do just about everything possible to obfuscate what's actually happening, and made it very hard to debug some problems that came up.  This also led to some minor changes in the build around when/how to run the SRA tests, but I think it's equivelant to how it's used now.

Having done it, I think it's working well.  Running a single test in IntelliJ is still very fast and gives useful results.  Running on the command line still takes several minutes, but is not any longer than before.  In retrospect I find the concerns around adding a few second of compile time somewhat laughable when the unit test suite takes ~4.5 minutes to run on a reasonable machine.

In other news, scalatest enables parallel testing, where it can run many suites in parallel to reduce runtime.  I had to disable this as there are tests in htsjdk that both fail, and **deadlock** if this option is enabled!!  If someone were willing to do some troubleshooting on that it could bring the test time way down.

The [second commit](https://github.com/samtools/htsjdk/pull/789/commits/85503ee9156e0fa7ba154d596b4248e81186ad8e#diff-0aaf596c2e8f7bfc61c4144488e74fd4) in that PR ports StringUtilTest to scala and adds some extra tests, to show what it's like writing tests in scalatest (at least, one style of it).  Generally speaking I think there are a few advantages:

1. I think the scalatest code is much more readable than the old TestNG code.  I've always particularly disliked the disconnect of tests and data-providers, which make it hard to follow, but that's just a small part of it.

2. You can use many of the syntax elements and library additions of scala to make writing tests much more rapid.

3. Because of the style of creating tests (`"x" should "y" in { z }`) and the extensive set of matchers (e.g. `foo.result should be <= 10`), when there are test failures you get massively more descriptive errors (e.g. `12 was not <= 10 in "description of your test"`).  

Clearly I'm on the side of doing this; I think the advantages outweigh the disadvantages.  What do others think?

**Comment by yfarjoun** [comment]
Given the PR #789 that @tfenne has provided, the remaining issues (as I see them) are as follows:

1. Check whether in IntelliJ a "find usages" action actually finds both java and scala tests (assuming a class has both) similar question when a test failes and dumps a stacktrace...is the jump from scala to java clear enough to use quickly in Intellij? 
2. A "style-guide" (first agreed upon here, and then perhaps moved to a readme.md) about coding and naming convention of TestFiles so that we can have both Scala and Java tests for any particular class
3. What actions in intelliJ does one need to perform in order to get this black-magic working? It's already difficult enough (I think) to get IntelliJ set-up correctly on the project, does it get harder with scala? Perhaps a short guide can be written...I want to make sure that the non-scala folks are kept on board and do not feel discouraged or left-out from this move.
4. Anything else? 

My suggestion for a start of a style guide (if folks want to edit we can move this to the wiki)
1. for Class Foo.java use tests FooTest.java or FooTest.scala or if you like/need FooJavaTest.java FooScalaTest.scala (so that a class can have both scala and java tests)
2. Avoid hard-core Scala in tests:
a. avoid high-order generic functions,
b. avoid implicit functions,  
3. use typedefs, named parameters and pattern-matching with abandon.

**Comment by magicDGS** [comment]
I would appreciate to have a Wiki regarding this, not only for implement but for understand what the tests are doing. I look at the PR and it seems quite similar to working with streams in Java 8, but I get a bit confused with the syntax of __shouldBe__ and other keywords; although they are verbose enough, I don't know if it could become more complicated when combined with others. For example, `shouldBe thrownBy` is pretty clear, but could be concatenated even more?

Overall, I can see the improvements of this in readability, but a Wiki guide will be very helpful just in case Scala newbies. For instance, how to include a new test case for a new functionality inside an already implemented one for do not reinvent the test in pure Java.

Thanks a lot for this!

**Comment by tfenne** [comment]
@magicDGS I'm happy to put a wiki page together assuming that we move forward with this.  In the meantime this page on scalatest is very useful for understanding the various kinds of "matcher" statements that you can use: http://www.scalatest.org/user_guide/using_matchers

**Comment by magicDGS** [comment]
Thanks @tfenne, looks great for me. From my side, +1 to the change!

**Comment by yfarjoun** [comment]
I've looked at FGbio's test suite which is quite extensive and runs in under 5 minutes, so I don't think that the time to test is a big issue. The only remaining issue, as I see it is that our agreement to limit the scala-ness of the tests might erode as more and more scala features creep into the code-base. To be clear, I am not opposed to that style of programing, but given that HTSJDK is a java project it would make it difficult for non-scala programers to read and extend scala tests. 

So I am in favor of this proposal and think that we should merge the PR that introduces scala tests to string util **and** write the wiki page that @tfenne offered to put together. In particular the page should specify which scala features are "allowed" in htsjdk testing.
